zcash_keys/src/keys.rs

```diff
@@ -928,6 +928,351 @@ impl UnifiedFullViewingKey {
     }
 }
 
+/// A [ZIP 316](https://zips.z.cash/zip-0316) unified incoming viewing key.
+#[derive(Clone, Debug)]
+pub struct UnifiedIncomingViewingKey {
+    #[cfg(feature = "transparent-inputs")]
+    transparent: Option<zcash_primitives::legacy::keys::ExternalIvk>,
+    #[cfg(feature = "sapling")]
+    sapling: Option<::sapling::zip32::IncomingViewingKey>,
+    #[cfg(feature = "orchard")]
+    orchard: Option<orchard::keys::IncomingViewingKey>,
+    /// Stores the unrecognized elements of the unified encoding.
+    unknown: Vec<(u32, Vec<u8>)>,
+}
+
+impl TryFrom<UnifiedFullViewingKey> for UnifiedIncomingViewingKey {
+    type Error = hdwallet::error::Error;
+
+    fn try_from(fvk: UnifiedFullViewingKey) -> Result<Self, Self::Error> {
+        Ok(UnifiedIncomingViewingKey {
+            #[cfg(feature = "transparent-inputs")]
+            transparent: fvk
+                .transparent
+                .map(|t| t.derive_external_ivk())
+                .transpose()?,
+            #[cfg(feature = "sapling")]
+            sapling: fvk.sapling.map(|s| s.to_external_ivk()),
+            #[cfg(feature = "orchard")]
+            orchard: fvk.orchard.map(|o| o.to_ivk(Scope::External)),
+            unknown: Vec::new(),
+        })
+    }
+}
+
+impl UnifiedIncomingViewingKey {
+    /// Construct a new unified incoming viewing key, if the required components are present.
+    pub fn new(
+        #[cfg(feature = "transparent-inputs")] transparent: Option<
+            zcash_primitives::legacy::keys::ExternalIvk,
+        >,
+        #[cfg(feature = "sapling")] sapling: Option<::sapling::zip32::IncomingViewingKey>,
+        #[cfg(feature = "orchard")] orchard: Option<orchard::keys::IncomingViewingKey>,
+        // TODO: Implement construction of UIVKs with metadata items.
+    ) -> Option<UnifiedIncomingViewingKey> {
+        #[cfg(feature = "orchard")]
+        let has_orchard = orchard.is_some();
+        #[cfg(not(feature = "orchard"))]
+        let has_orchard = false;
+        #[cfg(feature = "sapling")]
+        let has_sapling = sapling.is_some();
+        #[cfg(not(feature = "sapling"))]
+        let has_sapling = false;
+
+        if has_orchard || has_sapling {
+            Some(UnifiedIncomingViewingKey {
+                #[cfg(feature = "transparent-inputs")]
+                transparent,
+                #[cfg(feature = "sapling")]
+                sapling,
+                #[cfg(feature = "orchard")]
+                orchard,
+                // We don't allow constructing new UFVKs with unknown items, but we store
+                // this to allow parsing such UFVKs.
+                unknown: vec![],
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Parses a `UnifiedIncomingViewingKey` from its [ZIP 316] string encoding.
+    ///
+    /// [ZIP 316]: https://zips.z.cash/zip-0316
+    pub fn decode<P: consensus::Parameters>(params: &P, encoding: &str) -> Result<Self, String> {
```

Comment:
Instead of `String`, please create a new error enum.

Reply:
Also, instead of encoding and decoding functions that go to and from strings, these conversions should go to and from the `zcash_address::unified::Uivk` type. That then provides the conversion to and from strings.

Reply:
Ya, I feel string error types are gross. I only did this in the copy, but I was hoping you'd be ok with me changing it. Can I change in on the UFVK type as well?